File: sentimental-credit/credit.py
def main():
    # get input
    cardNumber = input("Enter card number:")

    # check input is valid
    if checkInput(cardNumber) == False:
        main() 

    length = len(cardNumber)

    # modulus regulates the card number length for the luhn algorithm
    modulus = int(length) % 2

    totalEvenNumbers = 0
    totalOddNumbers = 0
    totalCalculatedNumbers = 0

    counter = 0
    # iterate through the card number
    # calculates number according to luhn algoritm
    for i in cardNumber:
        if counter % 2 == modulus:
            tempEvenInt = int(i) * 2
            for j in str(tempEvenInt):
                totalEvenNumbers = totalEvenNumbers + int(j)
        else:
            totalOddNumbers += int(i)
        counter += 1

    totalCalculatedNumbers = totalOddNumbers + totalEvenNumbers

    if totalCalculatedNumbers % 10 == 0:
        checkCardType(cardNumber)
    else:
        print("INVALID")

# check input type


def checkInput(text):
    return text.strip().isdigit()

# check the card type according to card number and its length


def checkCardType(cardNumber):
    length = len(cardNumber)
    if length == 15 and cardNumber[0] == "3" and (cardNumber[1] == "4" or cardNumber[1] == "7"):
        print("AMEX")
    elif length == 16 and cardNumber[0] == "5" and (cardNumber[1] == "1" or cardNumber[1] == "2" or cardNumber[1] == "3" or cardNumber[1] == "4" or cardNumber[1] == "5"):
        print("MASTERCARD")
    elif (length == 13 or length == 16) and cardNumber[0] == "4":
        print("VISA")
    else:
        print("INVALID")


main()

File: credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <math.h>
#include <stdlib.h>

int main(void)
{
    long cardNumber = get_long("Enter card number:");
    bool result;
    // int lengthOfCardNumber = floor(log10(abs(cardNumber))) + 1;

    // convert long to string
    // char cardNumberStrTemp[16];
    char cardNumberStr[16];
    sprintf(cardNumberStr, "%ld", cardNumber);

    // reverse the card number

    printf("cardNumberStr %s\n", cardNumberStr);

    int lengthOfCardNumber = strlen(cardNumberStr);

    int modulus;

    // workaround for selecting number for luhn algorithm
    // if card length is even
    // modulus will be applied from first digit
    if (lengthOfCardNumber % 2 == 0)
        modulus = 0;
    else
        modulus = 1;

    int totalEvenNumbers = 0;
    int totalOddNumbers = 0;
    int totalCalculatedNumbers;
    for (int i = 0; i < lengthOfCardNumber; i++)
    {
        if ((i % 2) == modulus)
        {
            // workaround for multiplying the char
            // ###################################
            // to convert a 'char' to integer
            // we will subtract '0' from that 'char'
            // if we try converting directly string
            // we do not get the number itself
            // but we will get the ASCII value of char
            int tempEvenInt = cardNumberStr[i] - '0';

            // if any number multiplied by 2
            // result wont have more than two digits
            char tempEvenString[2];
            tempEvenInt = tempEvenInt * 2;

            // convert integer to string
            sprintf(tempEvenString, "%d", tempEvenInt);

            int lengthOfTempEvenNumber = strlen(tempEvenString);

            //calculate the total of elements from multiplied numbers
            for (int j = 0; j < lengthOfTempEvenNumber; j++)
            {
                // converting 'char' to 'int'
                totalEvenNumbers = totalEvenNumbers + tempEvenString[j] - '0';
            }
        }
        else
        {
            // converting 'char' to 'int'
            int tempOddInt = cardNumberStr[i] - '0';
            totalOddNumbers = totalOddNumbers + tempOddInt;
        }
    }

    totalCalculatedNumbers = totalOddNumbers + totalEvenNumbers;

    // check if modulus of total calculated number is 0
    // else return INVALID
    if ((totalCalculatedNumbers % 10) == 0)
    {
        result = true;
    }
    else
    {
        result = false;
        printf("INVALID\n");
        return 0;
    }

    // check for conditions for card types
    // else return INVALID
    if (result == true)
    {
        if (lengthOfCardNumber == 15 && cardNumberStr[0] == *"3" && (cardNumberStr[1] == *"4" || cardNumberStr[1] == *"7"))
        {
            printf("AMEX\n");
        }
        else if (lengthOfCardNumber == 16 && cardNumberStr[0] == *"5" && (cardNumberStr[1] == *"1" || cardNumberStr[1] == *"2" || cardNumberStr[1] == *"3" || cardNumberStr[1] == *"4" || cardNumberStr[1] == *"5"))
        {
            printf("MasterCard\n");
        }
        else if ((lengthOfCardNumber == 13 || lengthOfCardNumber == 16) && cardNumberStr[0] == *"4")
        {
            printf("Visa\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }

    return 0;
}